### PR TITLE
JAMES-3491 RFC-8887 implementation: JMAP transport over WebSocket

### PR DESCRIPTION
**Pull request by chibenwa**
TODO:

 - [x] Finish writing test and cleanup
 - [x] Investigate alternative testing libraries for WebSocket (reactor-netty is not an option)
 - [x] Check for debug statements in review

### CODE REVIEW

**Comment by rouazana** [comment]
started to try sttp for the client, it's almost ok: https://github.com/rouazana/james-project/tree/jmap-rfc-8887

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/WebSocketRoutes.scala [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        ProblemDetails(status = UNAUTHORIZED, detail = e.getMessage),
+        UNAUTHORIZED)
+    case e =>
+      LOGGER.error("Unexpected error upon WebSocket handshake request", e)
+      respondDetails(response,
+        ProblemDetails(status = INTERNAL_SERVER_ERROR, detail = e.getMessage),
+        INTERNAL_SERVER_ERROR)
+  }
+
+  private def asError(requestId: Option[RequestId])(throwable: Throwable): WebSocketError = throwable match {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/WebSocketRoutes.scala:125]: That's quite a bunches of duplicates.

I wonder if a `ProblemDetails.of(throwable)` would not allow mutualizing error handling between the HTTP and the websocket stack...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/WebSocketRoutes.scala [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        frame.content().readBytes(bytes)
+        new String(bytes, StandardCharsets.UTF_8)
+      })
+      .flatMap(handleClientMessages(session))
+      .onErrorResume(e => SMono.just(asError(None)(e)))
+      .map(ResponseSerializer.serialize)
+      .map(_.toString)
+      .flatMap(response => out.sendString(SMono.just(response), StandardCharsets.UTF_8))
+      .onErrorResume(e => {
+        e.printStackTrace()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/WebSocketRoutes.scala:87]: ouille

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+        |      "c1"
+        |    ]
+        |  ]
+        |}""".stripMargin)
+
+    await.until(() => client.receivedResponses.size() == 1)
+    assertThat(client.receivedResponses).hasSize(1)
+    assertThatJson(client.receivedResponses.get(0)).isEqualTo(
+      """
+        |{
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala:358]: We don't need these line breaks

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+      receivedResponses.add(s)
+    }
+
+    override def onClose(i: Int, s: String, b: Boolean): Unit = {
+      closeCode = Some(i)
+      closeString = Some(s)
+    }
+
+    override def onError(e: Exception): Unit = {
+      LOGGER.error("WebSocket error: " + e.getMessage)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala:74]: LOGGER.error("WebSocket error", e)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+  def setUp(server: GuiceJamesServer): Unit = {
+    server.getProbe(classOf[DataProbeImpl])
+      .fluent()
+      .addDomain(DOMAIN.asString())
+      .addUser(BOB.asString(), BOB_PASSWORD)
+  }
+
+  class ExampleClient(uri: URI) extends WebSocketClient(uri) {
+    val receivedResponses: util.LinkedList[String] = new util.LinkedList[String]()
+    var closeCode: Option[Integer] = None
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala:59]: We don't use that

**Comment by chibenwa** [comment]
Crazy compilation works locally

test this please

**Review by Arsnael — COMMENTED** [review]
Looks good to me so far :)

**Comment by chibenwa** [comment]
The compilation failure was caused by a rebase against master hence the force push

**Review by chibenwa — COMMENTED** [review]
:+1: looks great!

**Review by Arsnael — APPROVED** [review]

**Comment by LanKhuat** [comment]
`
[7e9eaf6876a8a161520fab29cd9d773ebbc21aea] [INFO] Running org.apache.james.mpt.DiscardProtocolTest
[7e9eaf6876a8a161520fab29cd9d773ebbc21aea] [ERROR] Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 61.052 s <<< FAILURE! - in org.apache.james.mpt.DiscardProtocolTest
[7e9eaf6876a8a161520fab29cd9d773ebbc21aea] [ERROR] testRecord  Time elapsed: 60.046 s  <<< ERROR!
[7e9eaf6876a8a161520fab29cd9d773ebbc21aea] java.lang.Exception: Timed out wait for be notified that read is complete
[7e9eaf6876a8a161520fab29cd9d773ebbc21aea] 	at org.apache.james.mpt.DiscardProtocolTest.testRecord(DiscardProtocolTest.java:84)
`

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**